modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskActivityServiceImpl.java

```diff
@@ -30,64 +37,120 @@ public TaskActivityServiceImpl(TaskActivitiesDataService taskActivitiesDataServi
     }
 
     @Override
-    public void addError(Task task, TaskHandlerException e, Map<String, Object> parameters) {
-        taskActivitiesDataService.create(new TaskActivity(e.getMessage(), e.getArgs(), task.getId(),
-                TaskActivityType.ERROR, ExceptionUtils.getStackTrace(e), parameters));
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public long addTaskStarted(Task task, Map<String, Object> parameters) {
+        int totalActions = task.getActions().size();
+        TaskActivity activity = taskActivitiesDataService.create(
+                new TaskActivity(TASK_IN_PROGRESS, Arrays.asList("0", String.valueOf(totalActions)), task.getId(), TaskActivityType.IN_PROGRESS, null, parameters, totalActions));
+        return activity.getId();
     }
 
     @Override
-    public void addSuccess(Task task) {
-        taskActivitiesDataService.create(new TaskActivity("task.success.ok", task.getId(),
-                TaskActivityType.SUCCESS));
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public boolean addSuccessfulExecution(Long activityId) {
+        TaskActivity activity = taskActivitiesDataService.findById(activityId);
+        if (activity == null) {
+            return false;
+        }
+
+        TaskExecutionProgress progress = activity.getTaskExecutionProgress();
+        progress.addSuccess();
+        boolean taskFinished = progress.getActionsSucceeded() == progress.getTotalActions();
+
+        if (taskFinished) {
+            activity.setActivityType(TaskActivityType.SUCCESS);
+            activity.setMessage(TASK_SUCCEEDED);
+            activity.getFields().clear();
```

Comment:
Is this because of a retry? Do we just keep updating the same activity? I would imagine I would have  a list of all failures, same as we have with regular retry, or am I missing something?